lake/scheduler.go

```diff
@@ -52,39 +54,48 @@ func (s *Scheduler) AddStats(stats zbuf.ScannerStats) {
 
 func (s *Scheduler) PullScanTask() (zbuf.PullerCloser, error) {
 	s.once.Do(func() {
-		go s.run()
+		s.run()
 	})
 	select {
 	case p := <-s.ch:
 		if p.Objects == nil {
-			return nil, <-s.done
+			return nil, s.group.Wait()
 		}
 		return s.newSortedScanner(p)
 	case <-s.ctx.Done():
-		return nil, <-s.done
+		return nil, s.group.Wait()
 	}
 }
 
 func (s *Scheduler) run() {
-	if err := ScanPartitions(s.ctx, s.snap, s.span, s.pool.Layout.Order, s.ch); err != nil {
-		s.done <- err
+	var ctx context.Context
+	s.group, ctx = errgroup.WithContext(s.ctx)
+	ch := s.ch
+	if s.index != nil {
+		ch = make(chan Partition)
```

Comment:
Give this channel the same amount of buffering as its companion?
```suggestion
		ch = make(chan Partition, cap(s.ch))
```

Reply:
Actually just this channel needs to be buffered since the object partition pass should be fast compared to the index and scan passes. Will change.